pkg/utils/managedresources/builder/managedresources.go

```diff
@@ -112,28 +115,94 @@ func (m *ManagedResource) DeletePersistentVolumeClaims(v bool) *ManagedResource
 	return m
 }
 
-// Reconcile creates or updates the ManagedResource.
+// Reconcile creates or updates the ManagedResource as well as marks all referenced secrets as garbage collectable.
 func (m *ManagedResource) Reconcile(ctx context.Context) error {
 	resource := &resourcesv1alpha1.ManagedResource{
 		ObjectMeta: metav1.ObjectMeta{Name: m.resource.Name, Namespace: m.resource.Namespace},
 	}
 
-	_, err := controllerutil.CreateOrUpdate(ctx, m.client, resource, func() error {
+	mutateFn := func(obj *resourcesv1alpha1.ManagedResource) {
 		for k, v := range m.labels {
-			metav1.SetMetaDataLabel(&resource.ObjectMeta, k, v)
+			metav1.SetMetaDataLabel(&obj.ObjectMeta, k, v)
 		}
 
 		for k, v := range m.annotations {
-			metav1.SetMetaDataAnnotation(&resource.ObjectMeta, k, v)
+			metav1.SetMetaDataAnnotation(&obj.ObjectMeta, k, v)
 		}
 
-		resource.Spec = m.resource.Spec
+		obj.Spec = m.resource.Spec
+
+		// the annotations should be injected after the spec is updated!
+		utilruntime.Must(references.InjectAnnotations(obj))
+	}
+
+	if err := m.client.Get(ctx, client.ObjectKeyFromObject(resource), resource); err != nil {
+		if apierrors.IsNotFound(err) {
+			// if the mr is not found just create it
+			mutateFn(resource)
+
+			// only mark the new secrets
+			if err := markSecretsAsGarbageCollectable(ctx, m.client, secretsFromRefs(resource)); err != nil {
+				return err
+			}
+
+			return m.client.Create(ctx, resource)
+		}
+		return err
+	}
+
+	// mark all old secrets as garbage collectable
```

Comment:
Can we add a TODO that this code is only relevant for migration and can be removed after a few releases, e.g. `v1.80`?

Reply:
Thinking about it again, we can't simply remove it (even in a high version) since there is no guaranteed migration step that processes the MRs of all affected clusters. Should we add such a step in the startup proceedure of Gardenlet, to handle all MRs of class `seed`?